### ISSUE TITLE
Provide a way to implement a MultiKueue support for the custom Jobs

### ISSUE THREAD

**Issue by tenzen-y** [kubernetes-sigs/kueue#2349]
**What would you like to be added**:
I would like to support the extension mechanism so that we can implement the MultiKueue controller for the custom / in-house Jobs. I'm thinking of exposing the `jobAdapter` interface and implementing a mechanism to add arbitrary objects to the `adapters` similar to the jobframework integration manager:

`adapters`: https://github.com/kubernetes-sigs/kueue/blob/e461fe0827786e5cd6f45ff6739ebeed9a700b05/pkg/controller/admissionchecks/multikueue/workload.go#L48-L51
`jobAdapters` interface: https://github.com/kubernetes-sigs/kueue/blob/e461fe0827786e5cd6f45ff6739ebeed9a700b05/pkg/controller/admissionchecks/multikueue/workload.go#L66-L81

`jobframework integration manager`: https://github.com/kubernetes-sigs/kueue/blob/e461fe0827786e5cd6f45ff6739ebeed9a700b05/pkg/controller/jobframework/integrationmanager.go

**Why is this needed**:
In many company, they have company specific CustomResources, and we should give a possibility to manage such CustomResources across multiple clusters.

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change
- [ ] Docs update

The artifacts should be linked in subsequent comments.

**Comment by alculquicondor** [comment]
+1
Is this something you want to give a first try to?

We've had requests to support kubeflow TFJobs

**Comment by tenzen-y** [comment]
> Is this something you want to give a first try to?

I will try this issue after I finish #2175.
So, feel free to assign this issue to anyone.

**Comment by dgrove-oss** [comment]
We should design this to support external integrations as well.

**Comment by tenzen-y** [comment]
> We should design this to support external integrations as well.

Yes, we can do it. But I think starting things as a minimum required is better. So, I want to take the external integrations as a separate issue.

**Comment by trasc** [comment]
For other jobs with built-in support in kueue it is fairly easy to do.

For fully external implementation it might be a bit harder, the multikueue admission check controller is composed of 3 object controllers (reconcilers for workload, admissionchecks and multikueue clusters)  and some long lived clients with watchers connected to the worker clusters. It's likely that the MulikueueCluster API needs to be modified to ensure that one multikueue cluster is only managed by one admission check controller.

For now I wold suggest to do the builtin jobs support and create a follow-up  for the external usage.

**Comment by tenzen-y** [comment]
> For other jobs with built-in support in kueue it is fairly easy to do.
> 
> For fully external implementation it might be a bit harder, the multikueue admission check controller is composed of 3 object controllers (reconcilers for workload, admissionchecks and multikueue clusters) and some long lived clients with watchers connected to the worker clusters. It's likely that the MulikueueCluster API needs to be modified to ensure that one multikueue cluster is only managed by one admission check controller.
> 
> For now I wold suggest to do the builtin jobs support and create a follow-up for the external usage.

@trasc This issue motivation is supporting in-house jobs. So, if you work only on built-in jobs like TFJob, could you open another issue?

**Comment by trasc** [comment]
> @trasc This issue motivation is supporting in-house/external jobs. So, if you work only on built-in jobs like TFJob, could you open another issue?

Done #2350

**Comment by tenzen-y** [comment]
> > @trasc This issue motivation is supporting in-house/external jobs. So, if you work only on built-in jobs like TFJob, could you open another issue?
> 
> Done #2350

@trasc Thank you. I'm guessing that we can work on this after #2350 is done because we will implement a mechanism to support more bilt-in Jobs in#2350, right?

**Comment by tenzen-y** [comment]
I synced supporting MultiKueue for the custom Jobs with @trasc.
In conclusion, we need to have 2 phases to support it in the following:

1. Move JobAdapters to JobFramework, and then expose the interface to other Go packages.
2. Implement a mechanism to specify AdmissionCheck ControllerName for the MultiKueue via JobFramwork.

In #2350, @trasc tries the first phase, and then he will try the second phase in this issue.

**Comment by tenzen-y** [comment]
@alculquicondor @trasc TBH, as an external contributor (non Googler), I'm feeling that the MultiKueue specifications is not clearing.
Because even though the current MultuKueue is significantly evolving different from the design phase, the KEP is too outdated (actually, we don't have the `MultiKueueConfig` resource in the implementations level): https://github.com/kubernetes-sigs/kueue/tree/main/keps/693-multikueue#design-details.

So, before we apply the something enhancements into the MultiKueue, shouldn't we update the KEP to align with the actual implementations and behavior?

After that, can we add new enhancements to the MultiKueue?

**Comment by trasc** [comment]
Hy @tenzen-y , indeed the KEP does not contain all the implementation details, but I don't think it should,  and I don't see any significant deviations from the design. 

To be specific  `MultiKueueConfig` is defined here: 
https://github.com/kubernetes-sigs/kueue/blob/8b217c67c6cfb2a588a90107a63d3cf19182b191/apis/kueue/v1alpha1/multikueue_types.go#L112-L118

and used:

https://github.com/search?q=repo%3Akubernetes-sigs%2Fkueue+%22MultiKueueConfig%22+language%3AGo&type=code

`<site>/docs/concepts/multikueue` and `<site>/docs/tasks/manage/setup_multikueue` are also providing additional information about MultiKueue.

**Comment by tenzen-y** [comment]
> To be specific MultiKueueConfig is defined here:

I meant the MultiKueueCluster.

> <site>/docs/concepts/multikueue and <site>/docs/tasks/manage/setup_multikueue are also providing additional information about MultiKueue.

That is user documentation, so I still think updating the MultiKueue proposal would be worth it since clarifying the internal mechanism would be worth it. Even if some maintainers step down from this project, this Kueue project continue to be the journey. Actually, I have some experience in the unclear internal specifications brought OSS projects bugs and challenging maintenance.

Additionally, the latest KEP (not outdated) allows us to clear the place where the rabbit hole is and could reduce the risks of generating traps by being reviewed by multiple contributors.

**Comment by mimowo** [comment]
Personally, I think this is a distinguished feature which justifies a dedicated KEP (even though I would be acceptive of MultiKueue KEP update). I would like to see the chosen approach outlined, and indicated alternatives, along with pros and cons. Additionally, the KEP process creates space for early feedback from users who are not code reviewers.

My ask is motivated by difficulty understanding design choices taken in the pending PRs (https://github.com/kubernetes-sigs/kueue/pull/2373 and https://github.com/kubernetes-sigs/kueue/pull/2405). Getting the decisions discussed and agreed early on could avoid the confusion, and potentially dropping a significant amount of code in case we decide to change the approach.

**Comment by trasc** [comment]
> I meant the MultiKueueCluster.

https://github.com/kubernetes-sigs/kueue/blob/8b217c67c6cfb2a588a90107a63d3cf19182b191/apis/kueue/v1alpha1/multikueue_types.go#L57-L85

**Comment by trasc** [comment]
> Additionally, the latest KEP (not outdated) allows us to clear the place where the rabbit hole is and could reduce the risks of generating traps by being reviewed by multiple contributors.

The KEP cannot cover all the implementation details, The details in all the KEPs are limited, take the provisioning admission check controller , a very similar component, there is no mentioning of it's internal components.

**Comment by alculquicondor** [comment]
I synced with @trasc offline.

We agreed that there are quite a few mini-components in MultiKueue that it is a bit hard to keep track of as an outsider. Traian will update the KEP with those design elements, spending about 1 paragraph for each mini-component.

I agree that most features don't require such level of detail, but the complexity in MultiKueue is a bit higher.

**Comment by trasc** [comment]
#2436 opened.

**Comment by alculquicondor** [comment]
#2373 looks like a simple refactoring overall to me. I'm ok merging it and have the documentation match those changes.

**Comment by mimowo** [comment]
> Actually, I synced with @trasc regarding to MultiKueue previous KEP, In conclusion, I agree with the current proposal, basically.

I'm sorry in advance if I ask stupid questions, but I still have some difficulty understanding the approach which I see continues to be implemented in https://github.com/kubernetes-sigs/kueue/pull/2458. 

Specifically:
- is it going to be a single AdmissionCheck per Job? In that case how a user can configure a single ClusterQueue to run multiple types of built-in jobs?
- will the new AdmissionChecks also run the built-in jobs, or the admins are expected to configure different admission checks for custom and built-in jobs?
- do we really need the new zoo of MultiKueue Admission checks? 

Can we at least consider a design where we still have a single MultiKueue admission check, but the custom jobs adapters are registered to it. I think it will make life on OnCall easier to have a single MutliKueue admission check that can be quickly identified by its name rather than a zoo of AdmissionChecks. It would also make admin users life easier to have a global configuration for the MK AC plugins, rather than doing this per ClusterQueue.

**Comment by trasc** [comment]
We are preparing a KEP for that.

**Comment by mimowo** [comment]
Ok great!

**Comment by tenzen-y** [comment]
> We are preparing a KEP for that.

That sounds great to me. I will review the submitted design soon.
Thank you.

**Comment by tenzen-y** [comment]
@mszadkow Is there any progressing on your KEP?

**Comment by mimowo** [comment]
We will probably postpone the KEP effort until after MK goes beta

**Comment by tenzen-y** [comment]
> We will probably postpone the KEP effort until after MK goes beta

That makes sense. The KEP could move forward during the MK beta graduation if @mszadkow has time.
But, we can prioritize the MK beta beta graduation rather than this feature.

**Comment by pramodbindal** [comment]
Hi,
We at RedHat are adding MultiKueue feature in Tekton Pipelines.

This feature will allow PipelineRuns to be distributed across cluster. But PipelineRun being external framework we are not able to admit PipelineRun jobs in MultiKueue.

```
status:
  admission:
    clusterQueue: cluster-queue
    podSetAssignments:
    - count: 1
      flavors:
        tekton.dev/pipelineruns: default-flavor
      name: pod-set-1
      resourceUsage:
        tekton.dev/pipelineruns: "1"
  admissionChecks:
  - lastTransitionTime: "2025-07-04T13:44:14Z"
    message: No multikueue adapter found for owner kind "tekton.dev/v1, Kind=PipelineRun"
    name: sample-multikueue
    state: Rejected
```

Please let me know the process I need to follow to be able to admit my PipelineRun Workloads to MultiKueue.

> _FYI: I am able to admit PipelineRun workloads to single cluster setup._

**Comment by tenzen-y** [comment]
IIRC, this is pending at https://github.com/kubernetes-sigs/kueue/pull/2458 due to deprioritization.
However, I think this is still a valid feature request.

/remove-lifecycle rotten

**Comment by mimowo** [comment]
FYI we have decided to go with the design merged in: https://github.com/kubernetes-sigs/kueue/pull/5981

~cc @Khurram~

**Comment by mimowo** [comment]
@khrm

**Comment by kannon92** [comment]
With https://github.com/kubernetes-sigs/kueue/pull/6760, I think we can mark this as done.

/close
